### PR TITLE
Internal Json : Add Utf8Span and Utf8Memory to Reader, Writer, Navigator + CosmosString

### PR DESCRIPTION
**Pull request by bchong95**
# Internal Json : Add Utf8Span and Utf8Memory to Reader, Writer, Navigator + CosmosString

## Description

Wires Utf8Span and Utf8Memory through the Json{Reader, Writer, Navigator} and CosmosString API. The benefits are that the new types are more reflective of the payload (they are utf8 byte sequences not just byte arrays) and enforce the contract. An additional benefit is that these types have syntactic sugar for comparing against Utf16Strings (regular strings). The names and API are similar to the new Utf8 APIs coming out in future .net core versions, so we can adopt them easily.  

So some before and after APIs

Reader Before:
`public override bool TryGetBufferedUtf8Value(out ReadOnlyMemory<byte> bufferedUtf8Value)`

Reader After:
`public override bool TryGetBufferedValue(out Utf8Memory bufferedValue)`

Writer Before:
`void WriteStringValue(ReadOnlySpan<byte> utf8StringValue);`

Writer After
`void WriteStringValue(Utf8Span value);`

Notice that we were able to get rid of Utf8 in the names since it's in the type. Also note that if we want to add Utf32 overload in the future, then we can do so through overloads instead of different names, since they won't both be ReadOnlyMemory<byte> anymore.

Internally I wired `UtfAllString`, which is a union of utf8 and utf16 string through `JsonBinaryEncoding` to cut down on the number of methods that are essentially copy and paste. 

I also broke up `JsonBinaryEncoding` into partial files, since the class is getting kind of large.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Json/JsonNavigator.cs [hunk 1]
```diff
@@ -68,9 +69,7 @@ protected JsonNavigator()
         public abstract Number64 GetNumberValue(IJsonNavigatorNode numberNode);
 
         /// <inheritdoc />
-        public abstract bool TryGetBufferedUtf8StringValue(
-            IJsonNavigatorNode stringNode,
-            out ReadOnlyMemory<byte> bufferedStringValue);
+        public abstract bool TryGetBufferedStringValue(IJsonNavigatorNode stringNode, out Utf8Memory bufferedStringValue);
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/JsonNavigator.cs:72]: Last I checked System.Text.Json doesn't support the extended types and newtonsoft's binary is BSON and not CosmosBinary. Also based on those performance tests binary beat out newtonsoft text, so I don't really see your point.

[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Json/JsonNavigator.cs:72]: System.Text.Json has dedicated Utf8 support, all writer/reader interactions are with Utf8Writer/Reader. When you say extended types, what do you mean? I can probably get the System.Text.Json team to answer it.

[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Json/JsonNavigator.cs:72]: This is the Reader exposed by System.Text.Json https://docs.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonreader?view=netcore-3.1

File: Microsoft.Azure.Cosmos/src/CosmosElements/CosmosString.cs [hunk 2]
```diff
@@ -23,7 +24,7 @@ protected CosmosString()
 
         public abstract string Value { get; }
 
-        public abstract bool TryGetBufferedUtf8Value(out ReadOnlyMemory<byte> bufferedUtf8Value);
+        public abstract bool TryGetBufferedValue(out Utf8Memory bufferedUtf8Value);
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/CosmosElements/CosmosString.cs:27]: >bufferedUtf8Value [](start = 64, length = 17)

value

File: Microsoft.Azure.Cosmos/src/Json/IReadOnlyJsonStringDictionary.cs [hunk 3]
```diff
@@ -14,8 +14,6 @@ namespace Microsoft.Azure.Cosmos.Json
 #endif
     interface IReadOnlyJsonStringDictionary
     {
-        bool TryGetStringAtIndex(int index, out string value);
-
-        bool TryGetUtf8StringAtIndex(int index, out ReadOnlyMemory<byte> value);
+        bool TryGetStringAtIndex(int index, out UtfAllString value);
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/IReadOnlyJsonStringDictionary.cs:17]: >bool TryGetStringAtIndex(int index, out UtfAllString value); [](start = 8, length = 60)

two overloads

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+    {
+        /// <summary>
+        /// Gets the string value from the binary reader.
+        /// </summary>
+        /// <param name="stringToken">The buffer that has the string.</param>
+        /// <param name="jsonStringDictionary">The JSON string dictionary.</param>
+        /// <returns>A string value from the binary reader.</returns>
+        public static string GetStringValue(
+            Utf8Memory stringToken,
+            JsonStringDictionary jsonStringDictionary)
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:21]: >JsonStringDictionary [](start = 12, length = 20)

IReadonly<T>

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+            if (JsonBinaryEncoding.TryGetBufferedStringValue(stringToken, jsonStringDictionary, out Utf8Memory bufferedUtf8StringValue))
+            {
+                result = bufferedUtf8StringValue.ToString();
+                return true;
+            }
+
+            result = default;
+            return false;
+        }
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:57]: as discussed, no need for this function

File: Microsoft.Azure.Cosmos/src/Json/JsonNavigator.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/JsonNavigator.cs:72]: extended types being int8, datetime, guid, etc.

File: Microsoft.Azure.Cosmos/src/Json/JsonNavigator.JsonBinaryNavigator.cs [hunk 6]
```diff
@@ -398,17 +404,17 @@ public override int GetObjectPropertyCount(IJsonNavigatorNode objectNode)
                 string propertyName,
                 out ObjectProperty objectProperty)
             {
-                ReadOnlyMemory<byte> buffer = JsonBinaryNavigator.GetNodeOfType(
+                _ = JsonBinaryNavigator.GetNodeOfType(
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/JsonNavigator.JsonBinaryNavigator.cs:407]: yes it applies validation. Using the discard makes it clear that the value isn't being used.

File: Microsoft.Azure.Cosmos/src/Json/JsonNavigator.JsonBinaryNavigator.cs [hunk 7]
```diff
@@ -398,17 +404,17 @@ public override int GetObjectPropertyCount(IJsonNavigatorNode objectNode)
...
                     objectNode);
 
-                ReadOnlySpan<byte> utf8StringPropertyName = Encoding.UTF8.GetBytes(propertyName);
+                Utf8Span utf8StringPropertyName = Utf8Span.TranscodeUtf16(propertyName);
                 foreach (ObjectProperty objectPropertyNode in this.GetObjectProperties(objectNode))
                 {
-                    if (this.TryGetBufferedUtf8StringValue(objectPropertyNode.NameNode, out ReadOnlyMemory<byte> bufferedUtf8StringValue))
+                    if (this.TryGetBufferedStringValue(objectPropertyNode.NameNode, out Utf8Memory bufferedUtf8StringValue))
                     {
                         // First try and see if we can avoid materializing the UTF16 string.
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/JsonNavigator.JsonBinaryNavigator.cs:416]: it avoids materializing the UTF16 string by comparing against the buffered utf8 string value.

File: Microsoft.Azure.Cosmos/src/Json/IReadOnlyJsonStringDictionary.cs (hunk 3 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/IReadOnlyJsonStringDictionary.cs:17]: as discussed will keep this

**Pull request merged**